File: providers/aws/src/test/java/datagen/services/ec2/model/SnapshotGenerator.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package datagen.services.ec2.model;

import datagen.BaseGenerator;
import datagen.CommonGenerator;
import software.amazon.awssdk.services.ec2.model.Snapshot;
import software.amazon.awssdk.services.ec2.model.SnapshotState;

import java.util.List;

public class SnapshotGenerator extends BaseGenerator {
    public static String snapshotId() {
        return String.format("snap-%s", faker.lorem().characters(10));
    }

    public static SnapshotState snapshotState() {
        return fromArray(SnapshotState.values());
    }

    public static List<Snapshot> snapshots() {
        return snapshots(faker.random().nextInt(1, 10));
    }

    public static List<Snapshot> snapshots(Integer n) {
        return listGenerator(n, SnapshotGenerator::snapshot);
    }

    public static Snapshot snapshot() {
        return Snapshot.builder()
                       .dataEncryptionKeyId(faker.lorem().word()) // TODO realistic value
                       .description(faker.lorem().sentence())
                       .encrypted(faker.random().nextBoolean())
                       .kmsKeyId(faker.lorem().word()) // TODO realistic value
                       .ownerId(CommonGenerator.accountId())
                       .progress(faker.random().nextInt(0, 100).toString())
                       .snapshotId(snapshotId())
                       .startTime(pastInstant())
                       .state(snapshotState())
                       .stateMessage(faker.lorem().sentence()) // TODO realistic value
                       .volumeId(VolumeGenerator.volumeId())
                       .volumeSize(faker.random().nextInt(1, 1000))
                       .ownerAlias(faker.lorem().word())
                       .tags(TagGenerator.tags())
                       .build();
    }
}

File: providers/aws/src/test/kotlin/datagen/services/ec2/model/SnapshotGenerator.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package datagen.services.ec2.model

import datagen.BaseGenerator
import datagen.CommonGenerator
import software.amazon.awssdk.services.ec2.model.Snapshot
import software.amazon.awssdk.services.ec2.model.SnapshotState

object SnapshotGenerator : BaseGenerator() {
    fun snapshotId(): String {
        return "snap-${faker.lorem().characters(30)}"
    }

    fun snapshotState(): SnapshotState {
        return valueFromArray(SnapshotState.values())
    }

    fun snapshots(n: Int? = null): List<Snapshot> {
        return listGenerator(n) { snapshot() }
    }

    fun snapshot(): Snapshot {
        return Snapshot.builder()
                .dataEncryptionKeyId(faker.lorem().word()) // TODO realistic value
                .description(faker.lorem().sentence())
                .encrypted(faker.random().nextBoolean())
                .kmsKeyId(faker.lorem().word()) // TODO realistic value
                .ownerId(CommonGenerator.accountId())
                .progress(faker.random().nextInt(0, 100).toString())
                .snapshotId(snapshotId())
                .startTime(pastInstant())
                .state(snapshotState())
                .stateMessage(faker.lorem().sentence()) // TODO realistic value
                .volumeId(VolumeGenerator.volumeId())
                .volumeSize(faker.random().nextInt(1, 1000))
                .ownerAlias(faker.lorem().word())
                .tags(TagGenerator.tags())
                .build()
    }
}
